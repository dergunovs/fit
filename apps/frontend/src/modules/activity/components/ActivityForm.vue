<template>
  <div v-if="isShowForm" data-test="activity-form-container">
    <h2>{{ t('activity.form') }}</h2>

    <UiFlex @submit.prevent="submit" tag="form" column gap="24" data-test="activity-form">
      <ActivityPotentialDuration :exercises="formData.exercises" data-test="activity-form-potential-duration" />

      <UiFlex justify="space-between">
        <UiButton @click="isShowTemplatesModal = true" isNarrow layout="secondary" data-test="activity-form-templates">
          {{ t('template.many') }}
        </UiButton>

        <UiButton
          @click="isShowCalendar = !isShowCalendar"
          :isDisabled="!isValid"
          isNarrow
          layout="secondary"
          data-test="activity-form-toggle-calendar"
        >
          {{ t('calendar') }}
        </UiButton>
      </UiFlex>

      <UiFlex v-if="isShowCalendar" column gap="16" data-test="activity-form-calendar-block">
        <UiFlex gap="4">
          <div>{{ dateScheduledText }}</div>

          <b v-if="formData.dateScheduled" data-test="activity-form-date-scheduled">
            {{ formatDate(formData.dateScheduled, locale) }}
          </b>
        </UiFlex>

        <UiButton
          @click="submit(true)"
          :isDisabled="!formData.dateScheduled || !isValid"
          data-test="activity-form-save-to-calendar"
        >
          {{ t('activity.saveToCalendar') }}
        </UiButton>

        <UiCalendar
          :minDate="new Date()"
          :lang="locale"
          @chooseDate="setScheduledDate"
          data-test="activity-form-calendar"
        />
      </UiFlex>

      <UiModal v-model="isShowTemplatesModal">
        <UiFlex column gap="32">
          <UiFlex column gap="16">
            <UiField :label="t('title')">
              <UiInput v-model="templateTitle" />
            </UiField>

            <UiButton
              @click="createTemplate"
              :isDisabled="templateTitle.length === 0 || formData.exercises.length === 0"
              layout="secondary"
            >
              {{ t('template.save') }}
            </UiButton>
          </UiFlex>

          <UiFlex v-if="user?.templates" column gap="16">
            <UiField :label="t('template.one')">
              <UiSelect
                v-model="currentTemplate"
                :lang="locale"
                :options="user.templates"
                data-test="activity-form-template-select"
              />
            </UiField>

            <UiButton @click="chooseTemplate" :isDisabled="!currentTemplate">{{ t('choose') }}</UiButton>
          </UiFlex>
        </UiFlex>
      </UiModal>

      <UiModal v-model="isShowExercisesModal" isScrollable data-test="activity-form-add-exercise-modal">
        <ExerciseChooseList
          v-if="exercises?.length"
          :exercises="exercises"
          @choose="addExercise"
          data-test="activity-form-exercise-choose-list"
        />
      </UiModal>

      <ExerciseChoosenList
        v-if="formData.exercises?.length"
        :choosenExercises="formData.exercises"
        @delete="deleteExercise"
        @createSet="createSet"
        data-test="activity-form-exercises-choosen"
      />

      <FormButtonsLayout isFixed>
        <UiButton @click="isShowExercisesModal = true" isNarrow data-test="activity-form-add-exercise">
          {{ t('exercise.add') }}
        </UiButton>

        <UiButton
          layout="accent"
          isNarrow
          :isDisabled="!isValid || isLoadingPost"
          type="submit"
          data-test="activity-form-submit"
        >
          {{ t('activity.start') }}
        </UiButton>
      </FormButtonsLayout>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue';
import { useRouter } from 'vue-router';
import { toast, UiButton, UiCalendar, UiFlex, UiModal, UiSelect, UiField, UiInput } from 'mhz-ui';
import { createTempId, deleteTempId, formatDate, useQueryClient } from 'mhz-helpers';
import {
  API_ACTIVITY,
  API_AUTH_GET,
  API_USER,
  IActivity,
  IExerciseChoosen,
  IUserTemplate,
  TPostActivityDTO,
} from 'fitness-tracker-contracts';

import ActivityPotentialDuration from '@/activity/components/ActivityPotentialDuration.vue';
import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import ExerciseChoosenList from '@/exercise/components/ExerciseChoosenList.vue';
import FormButtonsLayout from '@/common/components/FormButtonsLayout.vue';

import { exerciseService } from '@/exercise/services';
import { activityService } from '@/activity/services';
import { URL_ACTIVITY_EDIT } from '@/activity/constants';
import { URL_HOME } from '@/common/constants';
import { useTI18n } from '@/common/composables';
import { useAuthCheck } from '@/auth/composables';
import { generateActivityExercises } from '@/activity/helpers';
import { userService } from '@/user/services';

const router = useRouter();

const { t, locale } = useTI18n();
const queryClient = useQueryClient();
const { user } = useAuthCheck();

const formData = ref<IActivity>({
  exercises: [],
  isDone: false,
  dateScheduled: undefined,
});

const isShowTemplatesModal = ref(false);
const isShowExercisesModal = ref(false);
const isShowForm = ref(true);
const isShowCalendar = ref(false);

const currentTemplate = ref<IUserTemplate>();
const templateTitle = ref('');

const dateScheduledText = computed(() =>
  formData.value.dateScheduled ? t('activity.date') : t('activity.chooseDate')
);

const { data: exercises } = exerciseService.getAll();

function addExercise(exercise: IExerciseChoosen) {
  formData.value.exercises = formData.value.exercises?.length ? [...formData.value.exercises, exercise] : [exercise];

  isShowExercisesModal.value = false;
}

function deleteExercise(idToDelete: string) {
  formData.value.exercises = formData.value.exercises.filter((exercise) => exercise._id !== idToDelete);
}

function createSet() {
  const set = formData.value.exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  formData.value.exercises = [...formData.value.exercises, ...set];
}

function setScheduledDate(date: Date) {
  date.setHours(23, 59, 59);
  formData.value.dateScheduled = new Date(date);
}

function chooseTemplate() {
  isShowTemplatesModal.value = false;
  if (currentTemplate.value) formData.value.exercises = generateActivityExercises(currentTemplate.value.exercises);

  currentTemplate.value = undefined;
}

const { mutate: mutateUpdate } = userService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_USER] });
    await queryClient.refetchQueries({ queryKey: [API_AUTH_GET] });
    toast.success(t('template.added'));
  },
});

const { mutate: mutatePost, isPending: isLoadingPost } = activityService.create({
  onSuccess: async (activityId: TPostActivityDTO) => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY] });
    toast.success(formData.value.dateScheduled ? t('activity.savedToCalendar') : t('activity.started'));
    router.push(formData.value.dateScheduled ? URL_HOME : `${URL_ACTIVITY_EDIT}/${activityId}`);
  },
});

const isValid = computed(() => !!formData.value.exercises?.length);

function submit(isAddToCalendar?: boolean) {
  if (!isValid.value) return;

  isShowForm.value = false;
  if (!isAddToCalendar) formData.value.dateScheduled = undefined;
  formData.value.exercises = deleteTempId(formData.value.exercises);

  mutatePost(formData.value);
}

function createTemplate() {
  if (!formData.value.exercises) return;

  isShowTemplatesModal.value = false;

  const templateToCreate = { title: templateTitle.value, exercises: deleteTempId(formData.value.exercises) };

  const templates = user.value?.templates ? [...user.value.templates, templateToCreate] : [templateToCreate];

  if (user.value) mutateUpdate({ ...user.value, templates });

  currentTemplate.value = undefined;
  templateTitle.value = '';
}
</script>
