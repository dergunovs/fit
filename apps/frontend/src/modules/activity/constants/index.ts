import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
