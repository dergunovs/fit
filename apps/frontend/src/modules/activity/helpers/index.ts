import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
