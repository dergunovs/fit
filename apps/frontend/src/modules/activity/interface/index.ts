export interface IActivityCalendarEvent<T> {
  _id?: string;
  start: Date | null;
  end: Date | null;
  title: string;
  content: T[];
  color?: string;
}

export interface ICalendarDates {
  dateFrom: string;
  dateTo: string;
}

export interface ITimelineStep {
  left: number;
  right: number;
  type: 'rest' | 'exercise';
}
