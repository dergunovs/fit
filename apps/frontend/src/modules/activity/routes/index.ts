import {
  URL_ACTIVITY_CREATE,
  URL_ACTIVITY_EDIT,
  URL_ACTIVITY_ADMIN,
  URL_ACTIVITY_ADMIN_EDIT,
} from '@/activity/constants';

export const activityRoutes = [
  {
    path: URL_ACTIVITY_CREATE,
    name: 'ActivityCreate',
    component: () => import('@/activity/pages/ActivityCreatePage.vue'),
  },
  {
    path: `${URL_ACTIVITY_EDIT}/:activity`,
    name: 'ActivityPassing',
    component: () => import('@/activity/pages/ActivityPassingPage.vue'),
  },
  {
    path: URL_ACTIVITY_ADMIN,
    name: 'ActivityList',
    component: () => import('@/activity/pages/ActivityListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_ACTIVITY_ADMIN_EDIT}/:activity`,
    name: 'ActivityEdit',
    component: () => import('@/activity/pages/ActivityEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
