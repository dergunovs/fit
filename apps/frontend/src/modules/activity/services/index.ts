import { ComputedRef, Ref } from 'vue';
import {
  API_ACTIVITY,
  API_ACTIVITY_CALENDAR,
  API_ACTIVITY_LAST,
  API_ACTIVITY_STATISTICS,
  API_ACTIVITY_CHART,
  TActivityChartType,
  TGetActivitiesDTO,
  TGetActivitiesQueryDTO,
  TGetActivitiesCalendarDTO,
  TGetActivitiesCalendarQueryDTO,
  TGetActivitiesStatisticsDTO,
  TGetActivitiesStatisticsQueryDTO,
  TGetActivitiesChartDTO,
  TGetActivitiesChartQueryDTO,
  TGetActivityDTO,
  TGetActivityLastDTO,
  TPostActivityDTO,
  TPostActivityDataDTO,
  TUpdateActivityDTO,
  TUpdateActivityDataDTO,
  TDeleteActivityDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const activityService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_ACTIVITY, page],
      queryFn: async () => {
        const params: TGetActivitiesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetActivitiesDTO>(API_ACTIVITY, { params });

        return data;
      },
    });
  },

  getStatistics: (gap: number) => {
    return useQuery({
      queryKey: [API_ACTIVITY_STATISTICS],
      queryFn: async () => {
        const params: TGetActivitiesStatisticsQueryDTO = { gap };

        const { data } = await api.get<TGetActivitiesStatisticsDTO>(API_ACTIVITY_STATISTICS, { params });

        return data;
      },
    });
  },

  getCalendar: (options: object, dateFrom: Ref<string>, dateTo: Ref<string>) => {
    return useQuery({
      queryKey: [API_ACTIVITY_CALENDAR, dateFrom, dateTo],
      queryFn: async () => {
        const params: TGetActivitiesCalendarQueryDTO = { dateFrom: dateFrom.value, dateTo: dateTo.value };

        const { data } = await api.get<TGetActivitiesCalendarDTO>(API_ACTIVITY_CALENDAR, { params });

        return data;
      },
      ...options,
    });
  },

  getChart: (type: Ref<TActivityChartType>, isMonth: Ref<boolean>, locale: Ref<string>) => {
    return useQuery({
      queryKey: [API_ACTIVITY_CHART, type, isMonth, locale],
      queryFn: async () => {
        const params: TGetActivitiesChartQueryDTO = {
          type: type.value,
          month: isMonth.value.toString(),
          locale: locale.value,
        };

        const { data } = await api.get<TGetActivitiesChartDTO>(API_ACTIVITY_CHART, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_ACTIVITY, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetActivityDTO>(`${API_ACTIVITY}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  getLast: () => {
    return useQuery({
      queryKey: [API_ACTIVITY_LAST],
      queryFn: async () => {
        const { data } = await api.get<TGetActivityLastDTO>(API_ACTIVITY_LAST);

        return data.data;
      },
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY],
      mutationFn: async (formData: TPostActivityDataDTO) => {
        const { data } = await api.post<TPostActivityDTO>(API_ACTIVITY, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY],
      mutationFn: async (formData: TUpdateActivityDataDTO) => {
        const { data } = await api.patch<TUpdateActivityDTO>(`${API_ACTIVITY}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteActivityDTO>(`${API_ACTIVITY}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
