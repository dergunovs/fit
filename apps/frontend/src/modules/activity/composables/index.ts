import { ref } from 'vue';

import { ICalendarDates } from '@/activity/interface';

export function useActivityCalendar() {
  const dateFrom = ref('');
  const dateTo = ref('');

  const isDatesReady = ref(false);

  function updateDates(dates: ICalendarDates) {
    isDatesReady.value = true;

    dateFrom.value = dates.dateFrom;
    dateTo.value = dates.dateTo;
  }

  return { dateFrom, dateTo, isDatesReady, updateDates };
}
