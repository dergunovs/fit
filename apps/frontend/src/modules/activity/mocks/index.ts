import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
