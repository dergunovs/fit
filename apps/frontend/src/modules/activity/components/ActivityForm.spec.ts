import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest, formatDate } from 'mhz-helpers';
import { UiCalendar } from 'mhz-ui';
import { API_ACTIVITY } from 'fitness-tracker-contracts';

import ActivityForm from './ActivityForm.vue';
import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import FormButtonsLayout from '@/common/components/FormButtonsLayout.vue';

import { wrapperFactory } from '@/common/test';
import { mockOnSuccess, spyCreateActivity } from '@/activity/mocks';
import { spyRefetchQueries, spyRouterPush, spyToastSuccess } from '@/common/mocks';
import { EXERCISE_CHOOSEN_FIXTURE } from '@/exercise/fixtures';
import { URL_ACTIVITY_EDIT } from '@/activity/constants';
import { URL_HOME } from '@/common/constants';
import { spyGetExercisesAll } from '@/exercise/mocks';
import { spyUseAuthCheck } from '@/auth/mocks';

const form = dataTest('activity-form');
const formContainer = dataTest('activity-form-container');
const toggleCalendar = dataTest('activity-form-toggle-calendar');
const addExercise = dataTest('activity-form-add-exercise');
const exerciseChooseList = dataTest('activity-form-exercise-choose-list');
const calendarBlock = dataTest('activity-form-calendar-block');
const calendar = dataTest('activity-form-calendar');
const saveToCalendar = dataTest('activity-form-save-to-calendar');
const dateScheduled = dataTest('activity-form-date-scheduled');

let wrapper: VueWrapper<InstanceType<typeof ActivityForm>>;

beforeEach(() => {
  wrapper = wrapperFactory(ActivityForm, {}, { FormButtonsLayout });
});

enableAutoUnmount(afterEach);

describe('ActivityForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ActivityForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets initial data', async () => {
    expect(spyGetExercisesAll).toBeCalledTimes(1);
    expect(spyUseAuthCheck).toBeCalledTimes(1);
  });

  it('toggles calendar', async () => {
    expect(wrapper.find(calendar).exists()).toBe(false);

    await wrapper.find(toggleCalendar).trigger('click');

    expect(wrapper.find(calendarBlock).exists()).toBe(true);

    await wrapper.find(toggleCalendar).trigger('click');

    expect(wrapper.find(calendarBlock).exists()).toBe(false);
  });

  it('adds activity to calendar', async () => {
    wrapper.findComponent<typeof ExerciseChooseList>(exerciseChooseList).vm.$emit('choose', EXERCISE_CHOOSEN_FIXTURE);

    await nextTick();

    await wrapper.find(toggleCalendar).trigger('click');

    expect(wrapper.find(dateScheduled).exists()).toBe(false);

    const date = new Date();

    wrapper.findComponent<typeof UiCalendar>(calendar).vm.$emit('chooseDate', date);

    await nextTick();

    expect(wrapper.find(dateScheduled).text()).toBe(formatDate(date, 'ru'));

    await wrapper.find(saveToCalendar).trigger('click');

    date.setHours(23, 59, 59);

    expect(spyCreateActivity).toBeCalledTimes(1);

    const activityId = '123123';

    await mockOnSuccess.create?.(activityId);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_HOME);
  });

  it('creates activity', async () => {
    expect(spyCreateActivity).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    expect(wrapper.find(formContainer).exists()).toBe(true);

    await wrapper.find(addExercise).trigger('click');

    wrapper.findComponent<typeof ExerciseChooseList>(exerciseChooseList).vm.$emit('choose', EXERCISE_CHOOSEN_FIXTURE);

    await nextTick();

    await wrapper.find(form).trigger('submit');

    expect(wrapper.find(formContainer).exists()).toBe(false);

    expect(spyCreateActivity).toBeCalledTimes(1);

    const activityId = '123123';

    await mockOnSuccess.create?.(activityId);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(`${URL_ACTIVITY_EDIT}/${activityId}`);
  });
});
