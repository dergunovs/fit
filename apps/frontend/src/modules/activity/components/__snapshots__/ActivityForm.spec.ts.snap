// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityForm > matches snapshot 1`] = `
"<div data-test="activity-form-container">
  <h2>Сформировать занятие</h2>
  <div tag="form" column="" gap="24" data-test="activity-form">
    <activity-potential-duration-stub exercises="" data-test="activity-form-potential-duration"></activity-potential-duration-stub>
    <div justify="space-between"><button isnarrow="" layout="secondary" data-test="activity-form-templates">Шаблоны</button><button isdisabled="true" isnarrow="" layout="secondary" data-test="activity-form-toggle-calendar">Календарь</button></div>
    <!--v-if-->
    <div modelvalue="false">
      <div column="" gap="32">
        <div column="" gap="16">
          <fieldset label="Название"><input type="text" modelvalue=""></fieldset><button isdisabled="true" layout="secondary">Сохранить как шаблон</button>
        </div>
        <!--v-if-->
      </div>
    </div>
    <div modelvalue="false" isscrollable="" data-test="activity-form-add-exercise-modal">
      <exercise-choose-list-stub exercises="[object Object],[object Object],[object Object]" data-test="activity-form-exercise-choose-list"></exercise-choose-list-stub>
    </div>
    <!--v-if-->
    <div>
      <div class="_margin_871dab" data-test="form-buttons-layout-margin"></div>
      <div class="_buttons_871dab" data-fixed="true" data-test="form-buttons-layout-buttons"><button isnarrow="" data-test="activity-form-add-exercise">Добавить упражнение</button><button layout="accent" isnarrow="" isdisabled="true" type="submit" data-test="activity-form-submit">Начать занятие</button></div>
    </div>
  </div>
</div>"
`;
